docs/design/numeric_literals.md

```diff
@@ -0,0 +1,241 @@
+-   [Numeric literal semantics](#numeric-literal-semantics)
+-   [Open questions](#open-questions)
+    -   [Primitive types as code vs built-in](#primitive-types-as-code-vs-built-in)
+    -   [Syntax for wrapping operations](#syntax-for-wrapping-operations)
+    -   [Non-power-of-two sizes](#non-power-of-two-sizes)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Overview
+
+Numeric literals are used to define a numeric value in a program. This value can
+be of any type, such as an integral type or a real number type. Numeric literals
+have a type derived from their value. Two integer literals have the same type if
+and only if they represent the same integer. Two real number literals have the
+same type if and only if they represent the same real number.
+
+That is:
+
+-   For every integer, there is a type representing literals with that integer
+    value.
+-   For every rational number, there is a type representing literals with that
+    real value.
+-   The types for real numbers are distinct from the types for integers, even
+    for real numbers that represent integers. `var x: i32 = 1.0;` is invalid.
+
+Primitive operators are available between numeric literals, and produce values
+with numeric literal types. For example, the type of `1 + 2` is the same as the
+type of `3`.
+
+Numeric types can provide conversions to support initialization from numeric
+literals. Because the value of the literal is carried in the type, a type-level
+decision can be made as to whether the conversion is valid.
+
+The integer types defined in the standard library permit conversion from integer
+literal types whose values are representable in the integer type. The
+floating-point types defined in the Carbon library permit conversion from
+integer and rational literal types whose values are between the minimum and
+maximum finite value representable in the floating-point type.
+
+### Defined Types
+
+The following types are defined in the Carbon prelude:
+
+-   An arbitrary-precision integer type.
+
+    ```
+    class BigInt;
+    ```
+
+-   A rational type, parameterized by a type used for its numerator and
+    denominator.
+
+    ```
+    class Rational(T:! Type);
+    ```
+
+    The exact constraints on `T` are not yet decided.
+
+-   A type representing integer literals.
+
+    ```
+    class IntLiteral(N:! BigInt);
+    ```
+
+-   A type representing floating-point literals.
+
+        ```
+        class FloatLiteral(X:! Rational(BigInt));
+        ```
+
+    All of these types are usable during compilation. `BigInt` supports the same
+    operations as `Int(n)`. `Rational(T)` supports the same operations as
+    `Float(n)`.
+
+The types `IntLiteral(n)` and `FloatLiteral(x)` also support primitive integer
+and floating-point operations such as arithmetic and comparison, but these
+operations are typically heterogeneous: for example, an addition between
+`IntLiteral(n)` and `IntLiteral(m)` produces a value of type
+`IntLiteral(n + m)`.
+
```

Comment:
It would be good to have a TODO here to update this section once we have resolved how to reference things brought in by the prelude. My current understanding is that `BigInt`, `Rational`, `IntLiteral`, `FloatLiteral` will be accessed through a package prefix like `Carbon.BigInt` or `Core.BigInt`.